LandSandBoat/server
# Pet Level Restriction Interactions

**_I affirm:_**

## What does this pull request do?

Fixes https://github.com/HorizonFFXI/HorizonXI-Issues/issues/255
Pets will now change their level to reflect the master's level.

These changes primarily focus around handling pets with level restrictions.
- Add a `pet_level` column to `char_stats` that stores `petZoningInfo.petLevel`. This is currently only used for BST jugs. Initially there was some thought that all pets would stay at the level they were summoned at when not under level restriction but that only seems to be the case for jug pets.
- Refactor pet stat calculations to their own functions
- Handle level restrictions with pets including resetting them and recalculating stats

I'm working on an additional change to fix Automaton activation which does involve level restrictions but I decided it is a big enough change to warrant its own PR.

Thanks to Meloetta and @WinterSolstice8 for helping verify all of these interactions on retail!

## Steps to test these changes

Change your job to any pet job.
```
!changejob pup 75
!changejob smn 75
!changejob drg 75
!changejob bst 75
```
To apply and remove a level restriction use as well as check your pet stats.
```
/checkparam (on your pet)
!addeffect level_restriction 40
/checkparam (on your pet)
!deleffect level_restriction
/checkparam (on your pet)
```
